### ISSUE TITLE
Restart does not close old Electron instance

### ISSUE THREAD

**Issue by probablykasper** [electron-vite/vite-plugin-electron#251]
### Problem

My app has a `before-quit` handler which waits for the webview to be ready. But when the plugin makes Electron restart, the webview page/code is gone, so it can't tell the app to finish quitting. This means I end up with many Electron instances running at the same time

### Solution

Keep the Vite webpage loaded until the app quits

### Workaround

Disable restarting:

```js
onstart({ startup }) {
  if (process.electronApp) {
    console.log('\x1b[35mNot restarting\x1b[0m')
  } else {
    startup()
  }
}
```

**Comment by yejimeiming** [comment]
This problem happened before, and we solved it using `treeKillSync`.
Do you still have similar problems on you local machine?

https://github.com/electron-vite/vite-plugin-electron/blob/149c4a966530536050ac0890f46376a44dfe3027/src/index.ts#L166

https://github.com/electron-vite/vite-plugin-electron/blob/149c4a966530536050ac0890f46376a44dfe3027/src/utils.ts#L144-L150

**Comment by probablykasper** [comment]
Yes, I have the problem on the latest version (v0.28.7)

**Comment by probablykasper** [comment]
Yes I do have the problem on the latest version (v0.28.7).

However, it works if I call `treeKillSync` in my own Vite config, or even just `process.kill(process.electronApp.pid)`:
```js
onstart({ startup }) {
	if (process.electronApp) {
		treeKillSync(process.electronApp.pid)
		// or
		process.kill(process.electronApp.pid)
	}
	startup()
},
```

It would be neat if it did a graceful shutdown with `SIGTERM` by waiting for the child_process to exit

**Comment by yejimeiming** [comment]
Would you like to test the following changes locally?
https://github.com/electron-vite/vite-plugin-electron/blob/149c4a966530536050ac0890f46376a44dfe3027/src/utils.ts#L179

```diff
function killTree(tree: PidTree) {
  if (tree.children) {
    for (const child of tree.children) {
      killTree(child)
    }
  }

  try {
-   process.kill(tree.pid) // #214
+   process.kill(tree.pid, 'SIGKILL') // Force kill process
  } catch { /* empty */ }
}
```

**Comment by probablykasper** [comment]
That does work. Personally I would prefer waiting for the child process to exit though instead of force quitting

**Comment by probablykasper** [comment]
Here's a working solution without force quitting:
```ts
startup.exit = async () => {
  if (process.electronApp) {
    process.electronApp.removeAllListeners();
    process.kill(process.electronApp.pid, 'SIGTERM');
    await new Promise((resolve) => {
      process.electronApp.once("exit", resolve);
    })
  }
};
```

**Comment by yejimeiming** [comment]
> Here's a working solution without force quitting:
> 
> ```ts
> startup.exit = async () => {
>   if (process.electronApp) {
>     process.electronApp.removeAllListeners();
>     process.kill(process.electronApp.pid, 'SIGTERM');
>     await new Promise((resolve) => {
>       process.electronApp.once("exit", resolve);
>     })
>   }
> };
> ```

Does this wrok for you too? I would like to be like this.

**Comment by probablykasper** [comment]
Yes, it works for me. But I don't know how you want to incorporate `treeKillSync`

**Comment by yejimeiming** [comment]
Maybe you could try this for me, it should work fine.
`treeKillSync` is required, we need to take care of people using Windows systems.

```diff
startup.exit = async () => {
  if (process.electronApp) {
    process.electronApp.removeAllListeners()
    treeKillSync(process.electronApp.pid!)
+   await new Promise<number>((resolve) => {
+     process.electronApp.once('exit', resolve);
+   })
  }
}
```

**Comment by yejimeiming** [comment]
I will release `v0.28.8` in about an hour and a half.

**Comment by probablykasper** [comment]
That does not work unfortunately. It doesn't quit, and gives me this error:
```
[47765:0919/044144.460406:ERROR:gpu_process_host.cc(993)] GPU process exited unexpectedly: exit_code=15
[47765:0919/044144.468629:ERROR:network_service_instance_impl.cc(601)] Network service crashed, restarting service.
```

**Comment by yejimeiming** [comment]
Are you working on Mac OS?
It's worked on my Mac OS.

**Comment by yejimeiming** [comment]
<img width="908" alt="image" src="https://github.com/user-attachments/assets/79572786-b78e-45e1-9d51-0fbe390d7a8b">
The same error happened to me, but `vite-plugin-electron` still works fine. This error is thrown by Electron, it's not caused by us.

**Comment by probablykasper** [comment]
I am on macOS, yes. Are you testing with a `before-quit` handler like I use though?
https://github.com/probablykasper/ferrum/blob/7ac23dd2ca8fa397e51355aab6e0283f2e11be5a/src/electron/main.ts#L151-L168

**Comment by yejimeiming** [comment]
> I am on macOS, yes. Are you testing with a `before-quit` handler like I use though? https://github.com/probablykasper/ferrum/blob/7ac23dd2ca8fa397e51355aab6e0283f2e11be5a/src/electron/main.ts#L151-L168

This seems to be very tricky. `treeKillSync(pid)` directly kills the Renderer process(child process of Main process), causing the **ipc** handle you registered in `before-quit` to fail.
If we use `process.kill(pid)` instead of `treeKillSync(pid)`, this will cause some Windows users to be unable to quit the previous App process instance.

The current known way is to circumvent this problem by customizing the `onstart` option.

**Comment by yejimeiming** [comment]
@probablykasper Try this.
This may seem like a complete **hack**. But it can at least help us circumvent the problem we are facing now.

**Comment by probablykasper** [comment]
> If we use process.kill(pid) instead of treeKillSync(pid), this will cause some Windows users to be unable to quit the previous App process instance.

Do Wondows users have that issue even with the `.once('exit', resolve)`? Could be worth testing.

Seems weird if the tree-kill is the only way to quit the app 🤔

**Comment by yejimeiming** [comment]
Maybe the following code is better.

```js
startup.exit = async () => {
  if (process.electronApp) {
    await new Promise((resolve) => {
      process.electronApp.removeAllListeners()
      process.electronApp.once('exit', () => {
        treeKillSync(process.electronApp.pid)
        resolve()
      })
      process.kill(process.electronApp.pid)
    })
  }
}
```

**Comment by yejimeiming** [comment]
> Do Wondows users have that issue even with the .once('exit', resolve)? Could be worth testing.
>
> Seems weird if the tree-kill is the only way to quit the app 🤔

I created a `v0.28.9` branch which may fix this issue.
Would you like to build the `v0.28.9` branch locally and test if it works?

**Comment by probablykasper** [comment]
It does not work unfortunately. I made a reproduction repo here: https://github.com/probablykasper/vite-plugin-electron-bug

**Comment by yejimeiming** [comment]
> It does not work unfortunately. I made a reproduction repo here: https://github.com/probablykasper/vite-plugin-electron-bug

Private repo?

**Comment by probablykasper** [comment]
So sorry, public now

**Comment by kokororin** [comment]
I write a [module](https://www.npmjs.com/package/kill-electron) to terminate Electron zombie processes, which has been successfully used in my project for a long time. It has only been tested on Windows.

```ts
import electron from 'vite-plugin-electron/simple';
import { killElectron } from 'kill-electron';

export default defineConfig({
  plugins: [
    svelte({
      preprocess: vitePreprocess(),
    }),
    tailwindcss(),
    electron({
      onstart: args => {
        if (os.platform() === 'win32') {
          spawn('chcp', ['65001']);
        }
      
        if (process.env.VSCODE_DEBUG) {
          console.log(
            /* For `.vscode/.debug.script.mjs` */ '[startup] Electron App'
          );
        } else {
          void args.startup();
        }
        // Orphaned Electron processes can occur during debugging
        // To prevent system slowdowns, these processes need to be killed before each startup
        killElectron({
          zombieOnly: true
        });
      },
// ... rest code of vite config
```

**Comment by onburn** [comment]
vite-plugin-electron v0.28.8
vite v5.3.1

In mac, when I make modifications in the main process main.js, the BrowserWindow instance of electron will be re-created,
What should I do? 

```typescript
export default defineConfig({
  plugins: [
    vue(),
    vueJsx(),
    vueDevTools(),
    electron([
      {
        entry: "electron/main.js",
      },
      {
        entry: "electron/preload.js",
      },
    ]),
    electronRenderer(),
  ],
})
```

**Comment by hifron** [comment]
also present when stopped and are more instances which uses the same dev port...

Some of this https://github.com/vite-plugin repos may solve multiple instances or dev instances, but when there are multiple clients and users of Electron apps on Desktop, then this issue become something more when ports could be assigned for same port in different frameworks or different plugins or something different than Electron...

**Comment by hifron** [comment]
`man authbind`
_authbind  allows  a program which does not or should not run as root to
       bind to low-numbered ports in a controlled way._

if there is need for port open, find first unused port and use it. **In future** maybe to remember and register it to prevent another use(or cookie :), but **if it has some storage needs** then it is problem but some registration should solve it and it is already some bug here. And **maybe** make distinct  dev_ports and user_ports

In browser world there is PWA registered as chrome apps and in desktop world something for this needed port behaviour is maybe programatically in search of...

**Comment by hifron** [comment]
https://nuxt.com/docs/api/nuxt-config#loaders
https://nitro.build/guide
https://nitro.build/config#devproxy
serveStatic

if operating system does not support claiming port and there are also non-electron apps(like in MS Store), claiming first free port on specific range and setup some file config may not be enough... but that's also electron issue and not only...

**Comment by hifron** [comment]
https://chromestatus.com/feature/5133950203461632 not seemed as proper default handling issue...

**Comment by subframe7536** [comment]
Close for stale, feel free to reopen
